ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -54,26 +58,121 @@ import           Ouroboros.Consensus.HeaderValidation (HasAnnTip (..),
 import           Ouroboros.Consensus.Ledger.Query.Version
-import           Ouroboros.Consensus.Ledger.SupportsProtocol
 import qualified Ouroboros.Consensus.Ledger.Tables.DiffSeq as DS
-import           Ouroboros.Consensus.Ledger.Tables.Utils
 import           Ouroboros.Consensus.Node.NetworkProtocolVersion
                      (BlockNodeToClientVersion)
 import           Ouroboros.Consensus.Node.Serialisation
-                     (SerialiseNodeToClient (..), SerialiseResult (..))
+                     (SerialiseNodeToClient (..), SerialiseResult (..),
+                     SerialiseResult' (..))
 import           Ouroboros.Consensus.Storage.LedgerDB
 import           Ouroboros.Consensus.Storage.LedgerDB.API (LedgerDBView (..),
                      closeLedgerDBView)
 import qualified Ouroboros.Consensus.Storage.LedgerDB.BackingStore as BackingStore
 import           Ouroboros.Consensus.Storage.LedgerDB.DbChangelog
 import qualified Ouroboros.Consensus.Storage.LedgerDB.DbChangelog.Query as DbChangelog
-import           Ouroboros.Consensus.Util (ShowProxy (..), SomeSecond (..))
+import           Ouroboros.Consensus.Util (ShowProxy (..), Some (..))
 import           Ouroboros.Consensus.Util.DepPair
 import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.Singletons
 import           Ouroboros.Network.Block (HeaderHash, Point (..), StandardHash,
                      decodePoint, encodePoint)
 import           Ouroboros.Network.Protocol.LocalStateQuery.Type
-                     (ShowQuery (..))
+
+{-------------------------------------------------------------------------------
+  Footprints
+-------------------------------------------------------------------------------}
+
+-- | Queries on the local state might use ledger tables. This datatype (which
+-- will sometimes be concretized via @sing@) allows Consensus to categorize the
+-- queries.
+data QueryFootprint =
+    -- | The query doesn't need ledger tables, thus can be answered only with
+    -- the ledger state.
+    QFNone
+    -- | The query needs some tables, but doesn't need to traverse the whole
+    -- UTxO set.
+  | QFOne
+    -- | The query needs to traverse the whole UTxO set.
+  | QFAll
+
+data instance Sing (qf :: QueryFootprint) where
+  SQFNone :: Sing QFNone
+  SQFOne :: Sing QFOne
+  SQFAll :: Sing QFAll
+
+instance SingI QFNone where
+  sing = SQFNone
+instance SingI QFOne where
+  sing = SQFOne
+instance SingI QFAll where
+  sing = SQFAll
+
+type SomeBlockQuery :: (QueryFootprint -> Type -> Type) -> Type
+data SomeBlockQuery q =
+  forall fp result. SingI fp => SomeBlockQuery (q fp result)
+
+{-------------------------------------------------------------------------------
+  Block Queries
+-------------------------------------------------------------------------------}
+
+-- | Different queries supported by the ledger, indexed by the result type.
+type BlockQuery :: Type -> QueryFootprint ->  Type -> Type
```

Comment:
Might flipping the `blk` and `QueryFootPrint` arguments around simplify some parts of the code? It looks like we might not need `SerialiseResult'`, `SameDepIndex1` and `SomeBlockQuery` if we do

Reply:
This leads to complications on some instances (`fp0 !~ fp1`). It might be doable but I'm not doing it for now.